### ISSUE TITLE
Validation auc differs from sklearn roc_auc_score

### ISSUE THREAD

**Issue by travisbrady** [Microsoft/LightGBM#441]
## Environment info
* Operating System: Linux ip-172-31-0-177 4.4.0-65-generic #86-Ubuntu SMP Thu Feb 23 17:49:58 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
* CPU: Intel(R) Xeon(R) CPU E5-2666 v3 @ 2.90GHz (on AWS)
* C++/Python/R version: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609 and Anaconda Python 3.5

## Error Message:
No error, but LightGBM's auc differs greatly from sklearn's roc_auc_score. Is this expected? Are they definitely measuring the same thing?
Also, the problem is more pronounced the more rounds I train.

Example: 
Code:
```python
y_pred_proba = clf.predict_proba(df_test.values)
auc = roc_auc_score(y_test, y_pred_proba[:, 1])
```

Output with `n_estimators=50`, results fairly close:
```
[50]    valid_0's auc: 0.841734 valid_1's auc: 0.836171
AUC(macro): 0.8336487141008945
```

with `n_estimators=200`, results differ greatly:
```
[200]   valid_0's auc: 0.865004 valid_1's auc: 0.859736
AUC(macro): 0.7887123767354406
```

## Steps to reproduce

1. Fit binary LGBMClassifier with eval_set populated and eval_metric = 'auc'
2. Observe validation auc
3. Print sklearn roc_auc_score
4. Observe difference between lgbm and sklearn

**Comment by travisbrady** [comment]
Update:
Xgboost doesn't seem to exhibit this problem.
With the same code but with `XGBClassifier` dropped in I get (with `n_estimators=200`):

```
[199]   validation_0-auc:0.869524       validation_1-auc:0.863011
AUC(macro): 0.8630113982732611
```

**Comment by wxchan** [comment]
can you share your data? is there missing value (NaN) in your data?

**Comment by travisbrady** [comment]
@wxchan There are missing values (represented as NaN) in the features. Could that be the issue?
Or do you mean NaN's in the predictions themselves?

I'll see if I can share my data, may not be possible unfortunately.

**Comment by wxchan** [comment]
ok, you don't need to. I think it's caused by missing values, it's discussed in #437 .

**Comment by guolinke** [comment]
@travisbrady 
Can you try the latest code? I think I have fixed this in the latest code.

**Comment by travisbrady** [comment]
Thanks, @guolinke.
They're much closer now though not quite identical. Floating point issue maybe?

```
[200]   valid_0's auc: 0.870652 valid_1's auc: 0.864234
AUC(macro): 0.8631089788155538
```

**Comment by guolinke** [comment]
@travisbrady it is strange. Can you replace nan by yourself, and see what happen ?

**Comment by travisbrady** [comment]
Absolutely. What should I replace nan with? Is there some other missing value placeholder?

**Comment by guolinke** [comment]
use ```0``` .

**Comment by guolinke** [comment]
@travisbrady any updates ?

**Comment by travisbrady** [comment]
@guolinke Apologies. Haven't had the chance yet, checking momentarily.

**Comment by travisbrady** [comment]
Ok, I did fillna(0) with pandas and got the expected result:
```
[200]   valid_0's auc: 0.869543 valid_1's auc: 0.863152
AUC(macro): 0.8631514461711625
```

**Comment by guolinke** [comment]
ok. it seems fill 0 is working. 
@wxchan I have filled nan to 0 in c_api on this commit : https://github.com/Microsoft/LightGBM/commit/60c77487c15a69c9451be80fa9b15e987559a995  .
Why the pandas still need to fill 0 ? 

@travisbrady Did you have any NaN labels ?

**Comment by travisbrady** [comment]
No, all labels are 0 or 1.
```
print(y_train.value_counts())
print(y_test.value_counts())

False    3154430
Name: is_br, dtype: int64
False    342076
Name: is_br, dtype: int64
```

**Comment by goraj** [comment]
I have a similar issue when running cross validations using custom fold indices with lightgbm *2.0.11*.

Cross validation using a loop and the lightgbm.cv yield different mean AUCs.
The following example reproduces the behaviour:
```
import numpy as np
from sklearn.datasets import load_digits
from sklearn.metrics import roc_auc_score
import lightgbm as lgbm


if __name__ == '__main__':
    np.random.seed(4242)
    d = load_digits()


    xs = d['data']
    ys = d['target']
    indices = np.where((ys == 1) | (ys == 0))

    ys = ys[indices]
    xs = xs[indices]

    dset = lgbm.Dataset(xs, ys)

    model = None
    params = {
        'boosting_type': 'gbdt',
        'objective': 'binary',
        'learning_rate': 0.05,
        'num_leaves': 35,
        'metric': 'auc',
        'is_unbalance': True,
        'seed': 1024,
        'verbosity': -1
    }

    # create some folds
    niter = 100
    outer_folds = []
    aucv = []

    k = 5
    ss = len(xs)//k
    for i in np.arange(0, len(xs)-ss, ss):
        outer_folds.append((list(range(i, i+ss-10)), list(range(i+ss-10, i+ss))))

    # run a kfold cross validation using loop
    for i, (outer_trainset, outer_testset) in enumerate(outer_folds):
        trainset = lgbm.Dataset(xs[outer_trainset], ys[outer_trainset])
        model = lgbm.train(params, trainset, num_boost_round=niter)


        ydelta = model.predict(xs[outer_testset])
        aucv.append(roc_auc_score(ys[outer_testset], ydelta))

    # run a kfold cross validation using lightgbm cv
    r = lgbm.cv(params, folds=outer_folds, nfold=k, train_set=dset, shuffle=False, num_boost_round=niter)

    auc = r['auc-mean'][niter-1]
    stdev = r['auc-stdv'][niter-1]

    print('lightgbm cv auc: {:.4f} (+-{:.4f})'.format(auc, stdev))
    print('manual cv auc: {:.4f} (+-{:.4f})'.format(np.mean(aucv), np.std(aucv)))

```

```
lightgbm cv auc: 1.0000 (+-0.0000)
manual cv auc: 0.9792 (+-0.0361)

Process finished with exit code 0
```
